storage/gcp/issuers.go

```diff
@@ -57,35 +61,62 @@ func (s *IssuersStorage) keyToObjName(key []byte) string {
+// accross chains.
 func (s *IssuersStorage) AddIssuersIfNotExist(ctx context.Context, kv []storage.KV) error {
-	// We first try and see if this issuer cert has already been stored since reads
-	// are cheaper than writes.
-	// TODO(phboneff): add parallel operations
+	errG := errgroup.Group{}
 	for _, kv := range kv {
-		objName := s.keyToObjName(kv.K)
-		obj := s.bucket.Object(objName)
-
-		w := obj.If(gcs.Conditions{DoesNotExist: true}).NewWriter(ctx)
-		w.ContentType = s.contentType
-
-		if _, err := w.Write(kv.V); err != nil {
-			return fmt.Errorf("failed to write object %q to bucket %q: %w", objName, s.bucket.BucketName(), err)
-		}
-
-		if err := w.Close(); err != nil {
-			if ee, ok := err.(*googleapi.Error); ok && ee.Code == http.StatusPreconditionFailed {
-				for _, e := range ee.Errors {
-					if e.Reason == "conditionNotMet" {
-						klog.V(2).Infof("AddIssuersIfNotExist: object %q already exists in bucket %q, continuing", objName, s.bucket.BucketName())
-						return nil
+		errG.Go(func() error {
+			objName := s.keyToObjName(kv.K)
+			obj := s.bucket.Object(objName)
+
+			w := obj.If(gcs.Conditions{DoesNotExist: true}).NewWriter(ctx)
+			w.ContentType = s.contentType
+
+			if _, err := w.Write(kv.V); err != nil {
+				return fmt.Errorf("failed to write object %q to bucket %q: %w", objName, s.bucket.BucketName(), err)
+			}
+
+			if err := w.Close(); err != nil {
+				// If we run into a precondition failure error, check that the object
+				// which exists contains the same content that we want to write.
+				// If so, we can consider this write to be idempotently successful.
+				if ee, ok := err.(*googleapi.Error); ok && ee.Code == http.StatusPreconditionFailed {
+					existing, existingGen, err := s.getIssuer(ctx, objName)
+					if err != nil {
+						return fmt.Errorf("failed to fetch existing content for %q (@%d): %v", objName, existingGen, err)
 					}
+					if !bytes.Equal(existing, kv.V) {
+						klog.Errorf("Resource %q non-idempotent write:\n%s", objName, cmp.Diff(existing, kv.V))
+						return fmt.Errorf("precondition failed: resource content for %q differs from data to-be-written", objName)
+					}
+					klog.V(2).Infof("AddIssuersIfNotExist: object %q already exists in bucket %q, continuing", objName, s.bucket.BucketName())
+					return nil
 				}
+				return fmt.Errorf("failed to close write on %q: %v", objName, err)
 			}
+			klog.V(2).Infof("AddIssuersIfNotExist: added %q in bucket %q", objName, s.bucket.BucketName())
+			return nil
+		})
+	}
 
-			return fmt.Errorf("failed to close write on %q: %v", objName, err)
-		}
-
-		klog.V(2).Infof("AddIssuersIfNotExist: added %q in bucket %q", objName, s.bucket.BucketName())
+	if err := errG.Wait(); err != nil {
+		return err
 	}
+
 	return nil
 }
+
+// getIssuer returns the data and generation of the specified issuer, or an error.
+func (s *IssuersStorage) getIssuer(ctx context.Context, key string) ([]byte, int64, error) {
```

Comment:
Do you need to return the object generation from here?
I think you're only ever going to have a single generation, and aren't going to try to do anything like "overwrite iff object is generation X"?